finality-aleph/src/metrics/transaction_pool.rs

```diff
@@ -1,63 +1,137 @@
-use std::sync::Arc;
+use std::{
+    num::NonZeroUsize,
+    sync::Arc,
+    time::{Duration, Instant},
+};
 
-use futures::StreamExt;
-use sc_transaction_pool_api::{ImportNotificationStream, TransactionFor, TransactionPool};
-use sp_runtime::traits::Member;
+use lru::LruCache;
+use parking_lot::Mutex;
+use substrate_prometheus_endpoint::{
+    register, Counter, Histogram, HistogramOpts, PrometheusError, Registry, U64,
+};
 
-#[async_trait::async_trait]
-pub trait TransactionPoolInfoProvider {
-    type TxHash: Member + std::hash::Hash;
-    type Extrinsic: sp_runtime::traits::Extrinsic;
-    async fn next_transaction(&mut self) -> Option<Self::TxHash>;
+use crate::metrics::{exponential_buckets_two_sided, timing::Clock};
 
-    fn hash_of(&self, extrinsic: &Self::Extrinsic) -> Self::TxHash;
-}
+// Size of transaction cache: 32B (Hash) + 16B (Instant) * `100_000` is approximately 4.8MB
+const TRANSACTION_CACHE_SIZE: usize = 100_000;
+const BUCKETS_FACTOR: f64 = 1.4;
 
-pub struct TransactionPoolWrapper<T: TransactionPool> {
-    pool: Arc<T>,
-    import_notification_stream: ImportNotificationStream<T::Hash>,
+#[derive(Clone)]
+pub enum TransactionPoolMetrics<TxHash, C> {
+    Prometheus {
+        time_till_block_inclusion: Histogram,
+        transactions_not_seen_in_the_pool: Counter<U64>,
+        cache: Arc<Mutex<LruCache<TxHash, Instant>>>,
+        clock: C,
+    },
+    Noop,
 }
 
-impl<T: TransactionPool> TransactionPoolWrapper<T> {
-    pub fn new(pool: Arc<T>) -> Self {
-        Self {
-            pool: pool.clone(),
-            import_notification_stream: pool.import_notification_stream(),
-        }
-    }
-}
+impl<TxHash: std::hash::Hash + Eq, C: Clock> TransactionPoolMetrics<TxHash, C> {
+    pub fn new(registry: Option<&Registry>, clock: C) -> Result<Self, PrometheusError> {
+        let registry = match registry {
+            None => return Ok(Self::Noop),
+            Some(registry) => registry,
+        };
 
-#[async_trait::async_trait]
-impl<T: TransactionPool> TransactionPoolInfoProvider for TransactionPoolWrapper<T> {
-    type TxHash = T::Hash;
-    type Extrinsic = TransactionFor<T>;
+        Ok(Self::Prometheus {
+            time_till_block_inclusion: register(
+                Histogram::with_opts(
+                    HistogramOpts::new(
+                        "aleph_transaction_to_block_time",
+                        "Time from becoming ready in the pool to inclusion in some valid block.",
+                    )
+                    .buckets(exponential_buckets_two_sided(
+                        2000.0,
+                        BUCKETS_FACTOR,
+                        2,
+                        8,
+                    )?),
+                )?,
+                registry,
+            )?,
+            transactions_not_seen_in_the_pool: register(
+                Counter::new("aleph_transactions_not_seen_in_the_pool", "no help")?,
```

Comment:
Consider writing proper help messages after all maybe? Not a strict requirement, as I wouldn't quite know where to look for these other than in the code, but it's still nice to have stuff done properly.

There are some other places where it's set as "no help", this one just made me decide to write this comment after all. :P

Reply:
Added something